File: src/common/services/access-control.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, Repository } from 'typeorm';
import { AuthenticatedUser } from '../types/authenticated-user.type';
import { Department } from '../../modules/department/entities/department.entity';
import { Employee } from '../../modules/employee/entities/employee.entity';
import {
  EmployeeRole,
  hasMinimumRole,
} from '../../modules/employee/enums/employee-role.enum';

@Injectable()
export class AccessControlService {
  constructor(
    @InjectRepository(Department)
    private readonly departmentRepository: Repository<Department>,
    @InjectRepository(Employee)
    private readonly employeeRepository: Repository<Employee>,
  ) {}

  isAdmin(user: AuthenticatedUser): boolean {
    return user.role === EmployeeRole.ADMIN;
  }

  isProcurementManagerOrAbove(user: AuthenticatedUser): boolean {
    return hasMinimumRole(user.role, EmployeeRole.PROCUREMENT_MANAGER);
  }

  isDirectManagerOrAbove(user: AuthenticatedUser): boolean {
    return hasMinimumRole(user.role, EmployeeRole.DIRECT_MANAGER);
  }

  async getManagedDepartmentIds(userId: string): Promise<string[]> {
    const departments = await this.departmentRepository.find({
      where: { directManager: { id: userId } },
      select: { id: true },
    });
    return departments.map((department) => department.id);
  }

  async isEmployeeInManagedDepartments(
    managerId: string,
    employeeId: string,
  ): Promise<boolean> {
    const managedDepartmentIds = await this.getManagedDepartmentIds(managerId);
    if (!managedDepartmentIds.length) {
      return false;
    }

    const employee = await this.employeeRepository.findOne({
      where: { id: employeeId },
      relations: { department: true },
    });

    return Boolean(
      employee?.department?.id &&
      managedDepartmentIds.includes(employee.department.id),
    );
  }

  async getDepartmentEmployeeIds(departmentIds: string[]): Promise<string[]> {
    if (!departmentIds.length) {
      return [];
    }

    const employees = await this.employeeRepository.find({
      where: { department: { id: In(departmentIds) } },
      select: { id: true },
    });

    return employees.map((employee) => employee.id);
  }

  async findProcurementManager(): Promise<Employee | null> {
    return this.employeeRepository.findOne({
      where: { role: EmployeeRole.PROCUREMENT_MANAGER },
      order: { createdAt: 'ASC' },
    });
  }
}

File: src/EquipmentApi/Services/AccessControlService.cs
using EquipmentApi.Auth;
using EquipmentApi.Data;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class AccessControlService
{
    private readonly AppDbContext _db;

    public AccessControlService(AppDbContext db)
    {
        _db = db;
    }

    public bool IsAdmin(AuthenticatedUser user) => user.Role == EmployeeRole.Admin;

    public bool IsProcurementManagerOrAbove(AuthenticatedUser user) =>
        EmployeeRoleExtensions.HasMinimumRole(user.Role, EmployeeRole.ProcurementManager);

    public bool IsDirectManagerOrAbove(AuthenticatedUser user) =>
        EmployeeRoleExtensions.HasMinimumRole(user.Role, EmployeeRole.DirectManager);

    public async Task<List<Guid>> GetManagedDepartmentIdsAsync(Guid userId, CancellationToken cancellationToken = default)
    {
        return await _db.Departments
            .Where(department => department.DirectManagerId == userId)
            .Select(department => department.Id)
            .ToListAsync(cancellationToken);
    }

    public async Task<bool> IsEmployeeInManagedDepartmentsAsync(
        Guid managerId,
        Guid employeeId,
        CancellationToken cancellationToken = default)
    {
        var managedDepartmentIds = await GetManagedDepartmentIdsAsync(managerId, cancellationToken);
        if (managedDepartmentIds.Count == 0)
        {
            return false;
        }

        var employee = await _db.Employees
            .Include(e => e.Department)
            .FirstOrDefaultAsync(e => e.Id == employeeId, cancellationToken);

        return employee?.DepartmentId is not null
            && managedDepartmentIds.Contains(employee.DepartmentId.Value);
    }

    public async Task<List<Guid>> GetDepartmentEmployeeIdsAsync(
        IReadOnlyCollection<Guid> departmentIds,
        CancellationToken cancellationToken = default)
    {
        if (departmentIds.Count == 0)
        {
            return [];
        }

        return await _db.Employees
            .Where(employee => employee.DepartmentId != null && departmentIds.Contains(employee.DepartmentId.Value))
            .Select(employee => employee.Id)
            .ToListAsync(cancellationToken);
    }

    public Task<Employee?> FindProcurementManagerAsync(CancellationToken cancellationToken = default)
    {
        return _db.Employees
            .Where(employee => employee.Role == EmployeeRole.ProcurementManager)
            .OrderBy(employee => employee.CreatedAt)
            .FirstOrDefaultAsync(cancellationToken);
    }
}
